### ISSUE TITLE
Change colorspace to 4:2:2

### ISSUE THREAD

**Issue by a-tze** [voc/voctomix#33]
Rumors say that the current (default) config uses YUV420 as colorspace. This might be too lossy for some sophisticated scenarios. Therefore the default should be changed to YUV422.

Note: YUV444 does not make sense, since single-link HD SDI carries "only" YUV422 signals.

**Comment by MaZderMind** [comment]
The ffmpeg pix_fmt `yuv422p` is detected by gstreamer as `video/x-unknown, codec-id=(string)V_UNCOMPRESSED`.

The Decklink-Cards produce video with a pix_fmt of `uyvy422` which gstreamer detects as `video/x-raw, format=(string)UYVY` -- so this would be a good start.

Interestingly the `avenc_huffyuv` encoder (and fwiw also the `avenc_ffvhuff`) encoder mentioned in  https://github.com/voc/voctomix/issues/9#issuecomment-161053845 only support Y42B which may or may bot be just another name for the same thing. anyways, there will be at least one colorspace conversion here (in the ui-preview chain).

**Comment by MaZderMind** [comment]
Testing this in practice turned a small problem up:
after ~12 minutes, the input queue in ffmpeg's decklink code hits 1GB and ffmpeg starts to drop buffers. This can be reproduced with this command (on gstreamer 1.5.90):
````
ffmpeg -y \
  -f decklink \
  -i 'DeckLink Mini Recorder (1)@10' \
  -c:v rawvideo -c:a pcm_s16le \
  -f matroska \
  pipe: | gst-launch-1.0 fdsrc ! matroskademux name=demux \
    demux. ! queue ! video/x-raw,format=UYVY,width=1920,height=1080,framerate=25/1,pixel-aspect-ratio=1/1 ! intervideosink \
    demux. ! queue ! audio/x-raw,format=S16LE,channels=2,layout=interleaved,rate=48000 ! interaudiosink
````

changing `intervideosink` and `interaudiosink` to `fakesink`s fixes the problem, as well as changing from `UYVY`/`uyvy422` back to `I420`/`yuv420p`. For this reason on our production-system we reverted to the latter.

It's currently unclear weather this is a bug in gstreamer of ffmpeg or it it is me, holding it wrong.

**Comment by MaZderMind** [comment]
Ran another test on this tonight and it seems that leaving away the audio makes things work again:
````
ffmpeg -y \
  -f decklink \
  -i 'DeckLink Mini Recorder (1)@10' \
  -map 0:v \
  -c:v rawvideo \
  -f matroska \
  pipe: | gst-launch-1.0 fdsrc ! matroskademux name=demux \
    demux. ! queue ! video/x-raw,format=UYVY,width=1920,height=1080,framerate=25/1,pixel-aspect-ratio=1/1 ! intervideosink
````

**Comment by MaZderMind** [comment]
The issue is not focused on inter* elements, but is visible with all sync sinks:
````
identity instead of inter*
--------------------------
ffmpeg -y \
  -f decklink \
  -i 'DeckLink Mini Recorder (2)@10' \
  -c:v rawvideo -c:a pcm_s16le \
  -f matroska \
  pipe: | gst-launch-1.0 fdsrc ! matroskademux name=demux \
    demux. ! queue ! video/x-raw,format=UYVY,width=1920,height=1080,framerate=25/1,pixel-aspect-ratio=1/1 ! identity sync=true ! fakesink \
    demux. ! queue ! audio/x-raw,format=S16LE,channels=2,layout=interleaved,rate=48000 ! identity sync=true ! fakesink
````

It's also not depending on the amount of data sent, because with smaller video sizes it also fails:
````
scaled-down video
-----------------
ffmpeg -y \
  -f decklink \
  -i 'DeckLink Mini Recorder (2)@10' \
  -vf "scale=480:270" \
  -c:v rawvideo -c:a pcm_s16le \
  -f matroska \
  pipe: | gst-launch-1.0 fdsrc ! matroskademux name=demux \
    demux. ! queue ! video/x-raw,format=UYVY,width=480,height=270,framerate=25/1,pixel-aspect-ratio=1/1 ! identity sync=true ! fakesink \
    demux. ! queue ! audio/x-raw,format=S16LE,channels=2,layout=interleaved,rate=48000 ! identity sync=true ! fakesink
````

also setting the queue-sizes to maximum does not help:
````
queue-max
---------
ffmpeg -y \
  -f decklink \
  -i 'DeckLink Mini Recorder (1)@10' \
  -c:v rawvideo -c:a pcm_s16le \
  -f matroska \
  pipe: | gst-launch-1.0 fdsrc ! matroskademux name=demux \
    demux. ! queue max-size-buffers=0 max-size-bytes=0 max-size-time=0 ! video/x-raw,format=UYVY,width=1920,height=1080,framerate=25/1,pixel-aspect-ratio=1/1 ! identity sync=true ! fakesink \
    demux. ! queue max-size-buffers=0 max-size-bytes=0 max-size-time=0 ! audio/x-raw,format=S16LE,channels=2,layout=interleaved,rate=48000 ! identity sync=true ! fakesink
````

**Comment by MaZderMind** [comment]
So, I'm slowly getting desperate on this. During testing I found, that also the actual production scripts sometimes expose the very same problem (ie. when only starting core & one decklink source), while still using 420. So the issues documented above may not even be correlated to the colorspace.

**Comment by MaZderMind** [comment]
So, after all it seems the Reported "Buffer Overrun" seem to be totally unconnected to the Colorspace. Instead it seems that ffmpeg with the decklink input behaves differently, when no Source is connected -- and that was the reason for all the above problems.

**Comment by MaZderMind** [comment]
More tests show, that it's likely a combination.
running voctocore with one declink source in 422 worked for 42 hours, added a second one and both stopped working after ~1 hour.

My current guess is that using 422 upps the stream bandwidth from 593 MBit/s to 791 MBit/s and that triggers something in either ffmpeg or the core or the kernel handling the tcp connection. I'm currently trying to raise the fdsrc' buffer-size (which defaulted to 4k), which should reduce the number of syscalls and queue accesses.

**Comment by MaZderMind** [comment]
cleaned up the thread, removed false assumptions and added explanation of the issue.

**Comment by MaZderMind** [comment]
Example of how the problematic processes show up:
![screenshot from 2016-02-09 21-09-12](https://cloud.githubusercontent.com/assets/142237/12929226/e96ec06e-cf71-11e5-9e1d-b0f853decb47.png)
![screenshot from 2016-02-09 21-09-32](https://cloud.githubusercontent.com/assets/142237/12929225/e96c6d00-cf71-11e5-879e-a57ae320641a.png)

And a Video of how this looks in the GUI:
https://c3voc.mazdermind.de/permanent/issue-33-gui.mp4

**Comment by a-tze** [comment]
The screenshot shows a "-pix_fmt yuv420p" in the ffmpeg lines for capturing?

**Comment by MaZderMind** [comment]
@a-tze yes, true. The screenshot showed the problem when no source would be connected. In this case the colorspace would not matter.

**Comment by MaZderMind** [comment]
@a-tze is this still a thing after #143 or should we stay at 420?

**Comment by a-tze** [comment]
Well 422 would still be nice but I think most users (or even all) end up with 420 in their final encoded essences and it's "only" a quality thing.
